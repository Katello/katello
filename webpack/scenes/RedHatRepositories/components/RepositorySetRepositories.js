import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Alert, Spinner } from 'patternfly-react';
import { translate as __ } from 'foremanReact/common/I18n';

import loadRepositorySetRepos from '../../../redux/actions/RedHatRepositories/repositorySetRepositories';
import RepositorySetRepository from './RepositorySetRepository/';
import { yStream } from './RepositorySetRepositoriesHelpers';

export class RepositorySetRepositories extends Component {
  componentDidMount() {
    const { contentId, productId } = this.props;

    if (this.props.data.loading) {
      this.props.loadRepositorySetRepos(contentId, productId);
    }
  }

  sortedRepos = repos => [...repos.filter(({ enabled }) => !enabled)]
    .sort((repo1, repo2) => {
      const repo1YStream = yStream(repo1.releasever || '');
      const repo2YStream = yStream(repo2.releasever || '');

      if (repo1YStream < repo2YStream) {
        return -1;
      }

      if (repo2YStream < repo1YStream) {
        return 1;
      }

      if (repo1.arch === repo2.arch) {
        const repo1MajorMinor = repo1.releasever.split('.');
        const repo2MajorMinor = repo2.releasever.split('.');

        const repo1Major = parseInt(repo1MajorMinor[0], 10);
        const repo2Major = parseInt(repo2MajorMinor[0], 10);

        if (repo1Major === repo2Major) {
          const repo1Minor = parseInt(repo1MajorMinor[1], 10);
          const repo2Minor = parseInt(repo2MajorMinor[1], 10);

          if (repo1Minor === repo2Minor) {
            return 0;
          } return (repo1Minor > repo2Minor) ? -1 : 1;
        } return (repo1Major > repo2Major) ? -1 : 1;
      } return (repo1.arch > repo2.arch) ? -1 : 1;
    });

  render() {
    const { data, type } = this.props;

    if (data.error) {
      return (
        <Alert type="danger">
          <span>{data.error.displayMessage}</span>
        </Alert>
      );
    }

    const availableRepos = this.sortedRepos(data.repositories).map(repo => (
      <RepositorySetRepository key={repo.arch + repo.releasever} type={type} {...repo} />
    ));

    const repoMessage = (data.repositories.length > 0 && availableRepos.length === 0 ?
      __('All available architectures for this repo are enabled.') : __('No repositories available.'));

    return (
      <Spinner loading={data.loading}>
        {availableRepos.length ? availableRepos : <div>{repoMessage}</div>}
      </Spinner>
    );
  }
}

RepositorySetRepositories.propTypes = {
  loadRepositorySetRepos: PropTypes.func.isRequired,
  contentId: PropTypes.number.isRequired,
  productId: PropTypes.number.isRequired,
  type: PropTypes.string,
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    repositories: PropTypes.arrayOf(PropTypes.object),
    error: PropTypes.shape({
      displayMessage: PropTypes.string,
    }),
  }).isRequired,
};

RepositorySetRepositories.defaultProps = {
  type: '',
};

const mapStateToProps = (
  { katello: { redHatRepositories: { repositorySetRepositories } } },
  props,
) => ({
  data: repositorySetRepositories[props.contentId] || {
    loading: true,
    repositories: [],
    error: null,
  },
});

export default connect(mapStateToProps, {
  loadRepositorySetRepos,
})(RepositorySetRepositories);
