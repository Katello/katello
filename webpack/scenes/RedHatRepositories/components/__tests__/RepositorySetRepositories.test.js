import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { RepositorySetRepositories } from '../RepositorySetRepositories';

describe('RepositorySetRepositories Component', () => {
  it('should render with sorted repos', async () => {
    const shallowWrapper = shallow(<RepositorySetRepositories
      loadRepositorySetRepos={() => null}
      contentId={1}
      productId={2}
      data={{
        loading: false,
        repositories: [
          { arch: 'x86_64', releasever: '5.11' },
          { arch: 'x86_64', releasever: '7Server' },
          { arch: 'x86_64', releasever: '7.10' },
          { arch: 'x86_64', releasever: '7.1' },
          { arch: 'i386', releasever: '5.11' },
          { arch: 'i386', releasever: '5Workstation' },
          { arch: 'x86_64', releasever: '7.11' }],
      }}
    />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should render loading', async () => {
    const shallowWrapper = shallow(<RepositorySetRepositories
      loadRepositorySetRepos={() => null}
      contentId={1}
      productId={2}
      type="foo"
      data={{
        loading: true,
        repositories: [],
      }}
    />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should render with error', async () => {
    const shallowWrapper = shallow(<RepositorySetRepositories
      loadRepositorySetRepos={() => null}
      contentId={1}
      productId={2}
      type="foo"
      data={{
        loading: false,
        repositories: [],
        error: {
          displayMessage: 'some error',
        },
      }}
    />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });
});
