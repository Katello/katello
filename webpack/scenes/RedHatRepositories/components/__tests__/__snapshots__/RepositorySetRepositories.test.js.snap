// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositorySetRepositories Component should render loading 1`] = `
<Spinner
  className=""
  inline={false}
  inverse={false}
  loading={true}
  size="md"
>
  <div>
    No repositories available.
  </div>
</Spinner>
`;

exports[`RepositorySetRepositories Component should render with error 1`] = `
<Alert
  className=""
  onDismiss={null}
  type="danger"
>
  <span>
    some error
  </span>
</Alert>
`;

exports[`RepositorySetRepositories Component should render with sorted repos 1`] = `
<Spinner
  className=""
  inline={false}
  inverse={false}
  loading={false}
  size="md"
>
  <Connect(RepositorySetRepository)
    arch="x86_64"
    key="x86_647Server"
    releasever="7Server"
    type=""
  />
  <Connect(RepositorySetRepository)
    arch="i386"
    key="i3865Workstation"
    releasever="5Workstation"
    type=""
  />
  <Connect(RepositorySetRepository)
    arch="x86_64"
    key="x86_647.11"
    releasever="7.11"
    type=""
  />
  <Connect(RepositorySetRepository)
    arch="x86_64"
    key="x86_647.10"
    releasever="7.10"
    type=""
  />
  <Connect(RepositorySetRepository)
    arch="x86_64"
    key="x86_647.1"
    releasever="7.1"
    type=""
  />
  <Connect(RepositorySetRepository)
    arch="x86_64"
    key="x86_645.11"
    releasever="5.11"
    type=""
  />
  <Connect(RepositorySetRepository)
    arch="i386"
    key="i3865.11"
    releasever="5.11"
    type=""
  />
</Spinner>
`;
